src/radical/pilot/raptor/worker.py

```diff
@@ -195,73 +283,36 @@ def _call(self, data):
         args   = data.get('args',   [])
         kwargs = data.get('kwargs', {})
 
+        bak_stdout = sys.stdout
+        bak_stderr = sys.stderr
+
         try:
-            out = to_call(*args, **kwargs)
-            err = None
-            ret = 0
+            # redirect stdio to capture them during execution
+            sys.stdout = strout = StringIO()
+            sys.stderr = strerr = StringIO()
+
+            val = to_call(*args, **kwargs)
+            err = strout.getvalue()
+            ret = strerr.getvalue()
 
         except Exception as e:
             self._log.exception('_call failed: %s' % (data))
-            out = None
-            err = 'call failed: %s' % e
+            out = strout.getvalue()
+            err = strerr.getvalue() + ('\ncall failed: %s' % e)
             ret = 1
 
-        return out, err, ret
-
-
-    # --------------------------------------------------------------------------
-    #
-    # FIXME: an MPI call mode should be added.  That could work along these
-    #        lines of:
-    #
-    # --------------------------------------------------------------------------
-    #  def _mpi(self, data):
-    #
-    #      try:
-    #          cmd = rp.agent.launch_method.construct_command(data,
-    #                  executable=self.exe, args=data['func'])
-    #          out = rp.sh_callout(cmd)
-    #          err = None
-    #          ret = 0
-    #
-    #      except Exception as e:
-    #          self._log.exception('_mpi failed: %s' % (data))
-    #          out = None
-    #          err = 'mpi failed: %s' % e
-    #          ret = 1
-    #
-    #      return out, err, ret
-    # --------------------------------------------------------------------------
-    #
-    # For that to work we would need to be able to create a LM here, but ideally
-    # not replicate the work done in the agent executor.
-
-
-    # --------------------------------------------------------------------------
-    #
-    def _eval(self, data):
-        '''
-        We expect data to have a single entry: 'code', containing the Python
-        code to be eval'ed
-        '''
-
-        try:
-            out = eval(data['code'])
-            err = None
-            ret = 0
+        finally:
+            # restore stdio
+            sys.stdout = bak_stdout
+            sys.stderr = bak_stderr
 
-        except Exception as e:
-            self._log.exception('_eval failed: %s' % (data))
-            out = None
-            err = 'eval failed: %s' % e
-            ret = 1
 
-        return out, err, ret
+        return out, err, ret, val
```

Comment:
Assuming that the above `try...except` fails, what is the value of `val`?

Reply:
dito